index.js

```diff
@@ -835,6 +871,68 @@ class Batch {
   }
 }
 
+class Watcher extends EventEmitter {
+  constructor (bee, range) {
+    super()
+
+    this.bee = bee
+    this.core = bee.core
+
+    this.running = false
+    this.destroyed = false
+
+    this.range = range
+    this.latestDiff = this.bee.version
+    this.stream = null
+
+    this._onappend = this._onappend.bind(this)
+    this._onappend = debounceify(this._onappend)
+
+    // this.on('error', noop)
+  }
+
+  async _onappend () {
+    this.running = true
+
+    try {
+      await this._run()
+    } catch (err) {
+      this.emit('error', err)
+      this.destroy()
+    } finally {
+      this.running = false
+    }
+  }
+
+  async _run () {
+    if (this.destroyed) return
+
+    const snapshot = this.bee.snapshot()
```

Comment:
@HDegroote is this finally needed vs just storing the version?

Reply:
(tl;dr using snapshots avoids headaches)

Yes, I believe it is. Without a snapshot the following can happen:
- Suppose current version is 10, lastDiff 5
- Create a diffStream compared to 5 (asuming current version is 10)
- await the first entry of the stream
- Before the first entry resolves, a new element is added to the bee (so version 11)
- The first entry of the stream resolves, and the diffStream snaps from version 5->11

Without the snapshot it becomes pretty difficult to reason about. It's even tricky to figure out between which versions you have taken the diffStream (you can no longer do `    this.latestDiff = snapshot.version`. Doing `    this.latestDiff = this.bee.version` at the end could be wrong (bee can have updated), and doing it at the start could also be wrong as already illustrated.

From what I remember when investigating this, there's actually no safe way to know between which versions you have taken the diffStream of a non-snapshot bee, because it's possible for new entries to be added to the bee between the time the diffStream snaps and the time it returns control to the consumer for the first time.

It's a bit tricky to trigger in tests because the race conditions depend on the order on the event loop, which to some extend depends on CPU speed, but on my machine this triggers it:

First get rid of the snapshot:
`    const snapshot = this.bee // .snapshot() `

```
test.solo('multiple changes', async function (t) {
  t.plan(2)

  const db = create()

  const watcher = db.watch()
  t.teardown(() => watcher.destroy())

  watcher.on('change', function (leftVersion, rightVersion) {
    console.log(leftVersion, rightVersion)
  })

  await Promise.all([db.put('/a'), db.put('/b'), db.put('/c'), db.put('/d'), db.put('/e')])
  await sleep(500)
})
```
output:
```
3 1
4 3
6 5
```

so no trigger for 4->5

Reply:
@LuKks we should add an option to readstream, `head` thats the length at which it starts, then its trivial to not use a snap

Reply:
but we can do that in a follow up pr